### PR TITLE
Test pipeline parallelism works with DDP.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49577 Add benchmakr for torch.distributed.pipeline.sync.Pipe
* #48638 Improve documentation for pipeline parallelism.
* #48630 Cleanup APIs for pipeline parallelism.
* **#48470 Test pipeline parallelism works with DDP.**
* #49419 Use store based barrier in init_process_group.

Adding a unit test to test this works as expected. Although, this
doesn't work with other checkpointing modes of the pipe and checkpoint=never
needs to be set for this to work.

Differential Revision: [D25182668](https://our.internmc.facebook.com/intern/diff/D25182668/)

### CODE REVIEW

File: torch/testing/_internal/distributed/pipe_with_ddp_test.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+)
+from torch.distributed._pipeline.sync import Pipe
+import unittest
+
+class PipeWithDDPTest(RpcAgentTestFixture):
+    @property
+    def world_size(self) -> int:
+        return 4
+
+    @skip_if_lt_x_gpu(4)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/pipe_with_ddp_test.py:23]: IIRC, we will need to run ci-all to trigger multigpu tests to cover this

Comment (pritamdamania87) [reply torch/testing/_internal/distributed/pipe_with_ddp_test.py:23]: Tests are passing for multigpu: https://app.circleci.com/pipelines/github/pytorch/pytorch/250277/workflows/77bc3b3b-2a7c-46f7-b0ea-e06a496d1e4b/jobs/9579475, PR: https://github.com/pytorch/pytorch/pull/49302

File: torch/testing/_internal/distributed/pipe_with_ddp_test.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            )
+            model = Pipe(model, chunks=2, checkpoint=checkpoint)
+            model = DistributedDataParallel(model, process_group=pg)
+            out = model(torch.rand(16, 16).cuda(self.rank)).local_value()
+            out.sum().backward()
+
+            # Check grads
+            output = [torch.empty_like(fc1.weight.grad), torch.empty_like(fc1.weight.grad)]
+            pg.allgather(output, fc1.weight.grad)
+            torch.cuda.synchronize(self.rank)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/pipe_with_ddp_test.py:95]: curious, do we need to synchronize here? Will the compare op below be automatically inserted to the current stream?

File: torch/testing/_internal/distributed/pipe_with_ddp_test.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    requires_gloo,
+    requires_nccl,
+    skip_if_lt_x_gpu,
+    skip_if_rocm,
+)
+from torch.testing._internal.distributed.rpc.rpc_agent_test_fixture import (
+    RpcAgentTestFixture,
+)
+from torch.distributed._pipeline.sync import Pipe
+import unittest
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/pipe_with_ddp_test.py:16]: nit: shall we sort the import block a bit?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for adding this!

File: torch/testing/_internal/distributed/pipe_with_ddp_test.py (hunk 2 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/pipe_with_ddp_test.py]: Didn't need this and removed it.

**Pull request closed without merging**